src/classes/CW_AudienceGroupController.cls

```diff
@@ -0,0 +1,543 @@
+	{
+		if(this.setGroups!=null)
+			return this.setGroups.getHasPrevious();
+		else 
+			return null;
+	}
+
+	//return whether next page exists
+	public Boolean getHasNext()
+	{
+		if(this.setGroups!=null)
+			return this.setGroups.getHasNext();
+		else 
+			return null;
+	}
+
+	//return page number
+	public Integer getPageNumber()
+	{
+		if(this.setGroups!=null)
+			return this.setGroups.getPageNumber();
+		else 
+			return null;
+	}
+
+	//return total pages
+	Public Integer getTotalPages()
+	{
+		if (this.setGroups!=null)
+		{
+			pageSize = this.setGroups.getPageSize();
+			pages = totalSize/pageSize;
+		
+			return (Integer)pages.round(System.RoundingMode.CEILING);
+		}
+		else 
+			return null;
+	}
+	
+	//return count of selected items
+	public Integer getSelectedCount()
+	{
+		return this.selectedGroupIds.size();
+	}
+	
+	//return a set of control groups
+	public set<Id> getControlGroupIds()
+	{
+		controlGroupsIds = new Set<Id>();
+						  
+		for (CW_Audience_Group__c curr : audienceGroups)
+		{
+			if (curr.Control_Group__c!=null)
+				controlGroupsIds.add(curr.Control_Group__c);
+		}
+		
+		return controlGroupsIds;
+	}
+	
+	public String getDeleteMessage()
+	{
+		List<CW_Audience_Group__c> groupsToDelete = new List<CW_Audience_Group__c>();
+		List<CW_Audience_Group__c> groupsSelected = new List<CW_Audience_Group__c>();
+		List<CW_Audience_Group__c> selectedControlGroups = new List<CW_Audience_Group__c>();
+		
+		//get the set of control groups
+		controlGroupsIds = getControlGroupIds();
+		//list of groups selected for deletion
+		groupsSelected = [SELECT Id, Name, Is_Control_Group__c, 
+								 Description__c,
+								 Control_Group__c//, 
+								 //Number_of_Players_in_Group__c - TODO
+						  FROM CW_Audience_Group__c
+						  WHERE Id IN : this.selectedGroupIds];
+		for (CW_Audience_Group__c curr : groupsSelected)
+		{	
+			// check if the group selected is
+			if (controlGroupsIds.contains(curr.Id))
+			{	
+				//create list with all selected groups that are controlling groups as well
+				selectedControlGroups.add(curr);
+				//add the group name to the message
+				if (deleteMessage!='')
+					deleteMessage += ', ' + curr.Name ;
+				else deleteMessage += curr.Name ;
+			}  
+		}
+
+		//MORE than 1 group selected is a control group
+		if (selectedControlGroups.size()>1)
+		{
+			deleteMessage +=ConstantUtil.AUDIENCE_GROUP_MULTIPLE_CONTROL_GROUPS + System.Label.Audience_Group_Delete_Message;
+		} 
+		//only ONE group from the selected ones, is control group
+		else if (selectedControlGroups.size()==1)
+		{ 
+			deleteMessage += ConstantUtil.AUDIENCE_GROUP_SINGLE_CONTROL_GROUP + System.Label.Audience_Group_Delete_Message;
+		}
+		//NONE of the groups selected, is control group
+		else deleteMessage= ConstantUtil.AUDIENCE_GROUP_NO_CONTROL_GROUP;
+		
+		return deleteMessage;
+	}
+	
+	public String getCampaignName(Id campId)
+	{
+		String campName = [SELECT Name FROM CW_Campaign__c WHERE Id =: campId].Name;
```

Comment:
Problem: Unbound SOQL query. Reason: a SOQL query can return up to 50,000 records. If this number is exceeded a runtime exception is thrown. Suggestion: cap the max number of records returned adding a LIMIT clause to the query.